### PR TITLE
[Move] Fairy Lock

### PR DESCRIPTION
**Pull request by muscode13**
## [LOCALES PR](https://github.com/pagefaultgames/pokerogue-locales/pull/39)
I was told by a member of the translation team to merge the code in before the locales. If you'd like it the other way around, please let me know and I'll put this back in draft.

## What are the changes the user will see?
The move Fairy Lock will be implemented

## Why am I making these changes?
Fairy Lock in unimplemented

## What are the changes from a developer perspective?

- `FairyLockTag` - A new ArenaTag for FairyLock, it has the special text for its onAdd. This will also be referenced when checking the the pokemon is trapped
- In the pokemon.ts file, if the pokemon has the ArenaTagType.FAIRY_LOCK, they are considered trapped, in addition to the other conditions
- In the command-phase.ts file I added some logic for the text when attempting to switch out. It should prioritize the TrappedTag text if both exist. However if the TrappedTag doesn't exist and the Fairy Lock Tag does, then it will use the Fairy Lock text.

### Screenshots/Videos
Normal Battle:

https://github.com/user-attachments/assets/4c8d61df-65b5-4928-9b1a-d6696faef391

If a Pokémon that's on the field when Fairy Lock is used goes on to faint later in the same turn, the Pokémon that replaces it will still be unable to switch out in the following turn:

https://github.com/user-attachments/assets/a2427008-11b6-42b8-a553-78c02b2f0501

Fairy Lock does not prevent the effects of moves or items that force the target to switch, like U-turn:

https://github.com/user-attachments/assets/23023a2c-dcd9-41ee-9299-7671ac22b3a6

Fails if the effect is already active:

https://github.com/user-attachments/assets/f8ba5536-c68c-4ca7-b8af-3cecdd17f434

Not blocked by Protect:

https://github.com/user-attachments/assets/7e2581cf-200a-4048-9808-00a186710b35

Bypasses Substitute:

https://github.com/user-attachments/assets/6435a0c7-50c7-4177-ba41-7e6f9aa0b05e

Ghosts can escape:

https://github.com/user-attachments/assets/9639b31a-d976-4901-8619-335535932971

## How to test the changes?

- run `npm run test fairy_lock`
- Override to add Fairy Lock to Pokemon's movesets

## Checklist
- [x] **I'm using `beta` as my base branch**
- [x] There is no overlap with another PR?
- [x] The PR is self-contained and cannot be split into smaller PRs?
- [x] Have I provided a clear explanation of the changes?
- [x] Have I considered writing automated tests for the issue?
- [x] If I have text, did I make it translatable and add a key in the English locale file(s)?
- [x] Have I tested the changes (manually)?
    - [x] Are all unit tests still passing? (`npm run test`)
- [ ] Are the changes visual?
  - [x] Have I provided screenshots/videos of the changes?

### CODE REVIEW

File: src/test/moves/fairy_lock.test.ts
Comment (Tempo-anon) [inline src/test/moves/fairy_lock.test.ts:1]: Should add some tests for:
* Ghost types can still switch out of fairy lock
* Phazing moves (dtail, whilrwind, etc) will still switch out
* If a pokemon faints and is replaced the replacement should not be able to switch
* How this interacts with wimp out/emergency exit

Comment (muscode13) [reply src/test/moves/fairy_lock.test.ts:1]: Will do! Though the Wimp Out/EE test will have to wait until it gets merged into beta

File: src/phases/command-phase.ts [hunk 1]
```diff
@@ -228,23 +230,27 @@ export class CommandPhase extends FieldPhase {
...
             this.scene.ui.showText(
               i18next.t("battle:noEscapePokemon", {
-                pokemonName:  trapTag.sourceId && this.scene.getPokemonById(trapTag.sourceId) ? getPokemonNameWithAffix(this.scene.getPokemonById(trapTag.sourceId)!) : "",
-                moveName: trapTag.getMoveName(),
+                pokemonName: trapTag && trapTag.sourceId && this.scene.getPokemonById(trapTag.sourceId)
+                  ? getPokemonNameWithAffix(this.scene.getPokemonById(trapTag.sourceId)!)
+                  : fairyLockTag && fairyLockTag.sourceId && this.scene.getPokemonById(fairyLockTag.sourceId)
+                    ? getPokemonNameWithAffix(this.scene.getPokemonById(fairyLockTag.sourceId)!)
+                    : "",
+                moveName: trapTag ? trapTag.getMoveName() : fairyLockTag ? fairyLockTag.getMoveName() : "",
```
Comment (innerthunder) [inline src/phases/command-phase.ts:253]: Can this be broken down into if/elseif instead? Nested ternaries in general can be difficult to read.

File: src/test/moves/fairy_lock.test.ts [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+      .moveset([ Moves.FAIRY_LOCK, Moves.SPLASH ])
+      .ability(Abilities.BALL_FETCH)
+      .battleType("double")
+      .disableCrits()
+      .enemySpecies(Species.MAGIKARP)
+      .enemyAbility(Abilities.BALL_FETCH)
+      .enemyMoveset([ Moves.SPLASH, Moves.U_TURN ]);
+  });
+
+  it("Applies Fairy Lock tag for one turn, then apply Trapped tag for one turn", async () => {
```
Comment (innerthunder) [inline src/test/moves/fairy_lock.test.ts:36]: The test title here seems to be based on the old implementation?

File: src/test/moves/fairy_lock.test.ts [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+    expect(playerPokemon[1].isTrapped()).toEqual(true);
+    expect(enemyField[0].isTrapped()).toEqual(true);
+    expect(enemyField[1].isTrapped()).toEqual(true);
+
+    await game.toNextTurn();
+    expect(playerPokemon[0].isTrapped()).toEqual(false);
+    expect(playerPokemon[1].isTrapped()).toEqual(false);
+    expect(enemyField[0].isTrapped()).toEqual(false);
+    expect(enemyField[1].isTrapped()).toEqual(false);
+  });
```
Comment (innerthunder) [inline src/test/moves/fairy_lock.test.ts:66]:
```suggestion
  });

```
Give the tests some breathing room lol

File: src/test/moves/fairy_lock.test.ts [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+
+    game.move.select(Moves.SPLASH);
+    game.doSwitchPokemon(2);
+    await game.forceEnemyMove(Moves.SPLASH, 1);
+    await game.forceEnemyMove(Moves.SPLASH, 1);
+    await game.phaseInterceptor.to("BerryPhase");
+    await game.toNextTurn();
+
+    expect(game.scene.getPlayerField()[1].species.speciesId).not.toBe(Species.GENGAR);
+  });
```
Comment (innerthunder) [inline src/test/moves/fairy_lock.test.ts:92]:
```suggestion
  });

```

File: src/test/moves/fairy_lock.test.ts [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+    await game.forceEnemyMove(Moves.WHIRLWIND, 0);
+    game.doSelectPartyPokemon(2);
+    await game.forceEnemyMove(Moves.WHIRLWIND, 1);
+    game.doSelectPartyPokemon(2);
+    await game.phaseInterceptor.to("BerryPhase");
+    await game.toNextTurn();
+
+    expect(game.scene.getPlayerField()[0].species.speciesId).not.toBe(Species.KLEFKI);
+    expect(game.scene.getPlayerField()[1].species.speciesId).not.toBe(Species.TYRUNT);
+  });
```
Comment (innerthunder) [inline src/test/moves/fairy_lock.test.ts:118]:
```suggestion
  });

```

**Review by innerthunder — APPROVED** [review]

**Pull request merged**